### PR TITLE
feat(epic): error-handling consolidation — EPIC_ERROR_CODES + tightened parser + unified error type

### LINKED ISSUE #1101: epic-connector: unify operationOutcomeError() to throw EpicFhirError
## Context
Discovered during review of PR #1100.

PR #1100 introduces `EpicFhirError` and threads it through `requestWithRetry`, so 4xx/5xx HTTP errors now carry structured OperationOutcome data and downstream consumers can use `err.hasIssueCode("59022")`.

However, a second error-construction path remains as a plain `Error`:

```ts
// services/epic-connector/src/fhir-client.ts:481
function operationOutcomeError(outcome: OperationOutcome): Error {
  const issue = outcome.issue?.[0];
  const detail =
    issue?.diagnostics ?? issue?.details?.text ?? "no diagnostics";
  return new Error(
    `Epic returned OperationOutcome: ${issue?.code ?? "unknown"} - ${detail}`,
  );
}
```

This is called from `read()` (line 157), `createResource()` (line 245), and `updateResource()` (line 278) when Epic returns a 2xx HTTP status with an `OperationOutcome` resource body (the FHIR-spec way of indicating semantic failure on a successful HTTP call).

A successful Epic call returning an OperationOutcome with code `59022` would currently throw a plain `Error`, NOT an `EpicFhirError` — so the structured `err.hasIssueCode("59022")` check would fall through to the substring fallback. The whole point of PR #1100 is undermined for this path.

## Proposal
Refactor `operationOutcomeError()` to return `EpicFhirError`:

```ts
function operationOutcomeError(outcome: OperationOutcome, status = 200, statusText = "OK"): EpicFhirError {
  const issue = outcome.issue?.[0];
  const detail =
    issue?.diagnostics ?? issue?.details?.text ?? "no diagnostics";
  return new EpicFhirError(
    `Epic returned OperationOutcome: ${issue?.code ?? "unknown"} - ${detail}`,
    status,
    statusText,
    JSON.stringify(outcome),
    outcome,
  );
}
```

Then `read()`/`createResource()`/`updateResource()` callers benefit from `hasIssueCode()` without needing the substring fallback.

## Acceptance
- [ ] `operationOutcomeError()` returns `EpicFhirError` with `operationOutcome` populated
- [ ] `read()`/`createResource()`/`updateResource()` propagate the typed error unchanged
- [ ] Test: `client.read()` against a 200 + OperationOutcome body throws `EpicFhirError` with `hasIssueCode("59022") === true`
- [ ] Sync-jobs substring fallback for the `"Epic returned OperationOutcome:"` shape can then be deleted (or kept one release with a `log.warn` flagging that the structured path didn't fire)

## Refs
- PR #1100
- `services/epic-connector/src/fhir-client.ts` lines 157, 245, 278, 481

### LINKED ISSUE #1102: epic-connector: extract EPIC_ERROR_CODES registry for structured failure discrimination
## Context
Discovered during review of PR #1100.

PR #1100 lands a single Epic-code constant:

```ts
// services/epic-connector/src/sync/sync-jobs.ts
const EPIC_UNAUTHORIZED_SUB_RESOURCE_CODE = "59022";
```

`EpicFhirError`'s docstring references `59108` ("A required element is missing") as the next code the connector will likely need to discriminate, and Epic publishes a fuller error-code table. Future code-based checks should look up by symbolic name, not duplicate string literals across modules.

## Proposal
Extract a single source of truth in `services/epic-connector/src/epic-error-codes.ts`:

```ts
/** Epic FHIR error codes the connector reasons about structurally. */
export const EPIC_ERROR_CODES = {
  /** Combination of parameters is not valid for any authorized sub-resource.
   *  Soft-skip — registered app lacks scope for this filter combo. */
  UNAUTHORIZED_SUB_RESOURCE: "59022",
  /** A required element is missing. Surface — request-shape bug. */
  MISSING_REQUIRED_ELEMENT: "59108",
} as const;

export type EpicErrorCode =
  (typeof EPIC_ERROR_CODES)[keyof typeof EPIC_ERROR_CODES];
```

Then in `sync-jobs.ts`:
```ts
if (err.hasIssueCode(EPIC_ERROR_CODES.UNAUTHORIZED_SUB_RESOURCE)) return true;
```

Optional: re-type `EpicFhirError.hasIssueCode` to accept `EpicErrorCode | string` so symbolic names are autocomplete-discoverable.

## Acceptance
- [ ] `EPIC_ERROR_CODES` registry exported with at least `UNAUTHORIZED_SUB_RESOURCE` and `MISSING_REQUIRED_ELEMENT`
- [ ] `sync-jobs.ts` uses the symbolic constant
- [ ] Each entry has a docstring explaining the Epic semantic + the connector's policy (soft-skip vs surface)

## Refs
- PR #1100
- `services/epic-connector/src/sync/sync-jobs.ts` (current `EPIC_UNAUTHORIZED_SUB_RESOURCE_CODE`)
- `services/epic-connector/src/fhir-client.ts` (`EpicFhirError` docstring referencing 59022/59108)

### LINKED ISSUE #1103: epic-connector: tests for tryParseOperationOutcome edge cases (non-JSON, non-object, wrong resourceType)
## Context
Discovered during review of PR #1100.

`tryParseOperationOutcome()` in `services/epic-connector/src/fhir-client.ts` is the trust boundary for promoting raw Epic response bodies into typed `OperationOutcome` data that downstream callers `hasIssueCode()` against. Today only two integration cases exercise it:

- Happy path: valid OperationOutcome JSON (covered by the new `EpicFhirError with parsed OperationOutcome on 4xx` test)
- HTML/plain text body: returns `undefined` (covered)

Gaps:

1. Valid JSON but NOT an object (`"hello"`, `42`, `[1,2,3]`, `null`)
2. Valid JSON object without `resourceType`
3. Valid JSON object with `resourceType !== "OperationOutcome"` (e.g. `"Patient"` body returned by mistake)
4. Valid OperationOutcome with `issue` missing or non-array (current `operationOutcomeHasIssueCode` does `for (const issue of outcome.issue)` — `undefined.issue` would throw if the parser accepts a doc without `issue`)
5. Valid OperationOutcome where `details.coding` is missing entirely (covered by the `?? []` default, but no negative test locks it in)

A future refactor that, say, loosens the type narrowing or drops the `resourceType` check could silently start accepting non-OperationOutcome payloads as if they were OperationOutcomes.

## Acceptance
- [ ] Direct unit tests for `tryParseOperationOutcome` covering the 5 cases above
- [ ] Test that an EpicFhirError constructed with an OperationOutcome that has `issue: []` returns `false` from `hasIssueCode("59022")` (not throws)
- [ ] Consider tightening the parser to also require `Array.isArray(parsed.issue)` before returning

## Note on prototype pollution
`JSON.parse` itself does not assign to `Object.prototype` so prototype pollution isn't a direct risk, but the type assertion `parsed as OperationOutcome` is unchecked beyond `resourceType`. Worth at least asserting the shape minimally before trusting it.

## Refs
- PR #1100
- `services/epic-connector/src/fhir-client.ts` (`tryParseOperationOutcome`)
- `services/epic-connector/src/fhir-types.ts` (`operationOutcomeHasIssueCode`)

### LINKED ISSUE #1104: epic-connector: tighten OperationOutcomeIssueCoding.code typing (FHIR R4 requires code when coding present)
## Context
Discovered during review of PR #1100.

`OperationOutcomeIssueCoding` in `services/epic-connector/src/fhir-types.ts` is currently:

```ts
export interface OperationOutcomeIssueCoding {
  system?: string;
  code?: string;
  display?: string;
}
```

FHIR R4 spec (`Coding` datatype, http://hl7.org/fhir/R4/datatypes.html#Coding) does not strictly require `code` (it's `0..1` in the structure definition), but in practice if a Coding has no `code` it carries no useful information for structured discrimination. Today `hasIssueCode` will silently return `false` for a coding entry where `code` is undefined, which is correct behavior but masks the upstream data quality issue.

## Proposal
Two options:

1. **Tighten the type** to require `code`:
   ```ts
   export interface OperationOutcomeIssueCoding {
     system?: string;
     code: string;  // required by Epic convention
     display?: string;
   }
   ```
   Trade-off: more honest typing of what Epic actually returns; forces a guard in `tryParseOperationOutcome`.

2. **Keep optional but log when missing**:
   ```ts
   if (!c.code) log.debug({ system: c.system }, "OperationOutcome coding missing code");
   ```
   Trade-off: type stays spec-permissive; surfaces unexpected payloads as observable warnings.

Option 1 is preferred for the Epic-specific call site since Epic always emits a code in practice; if a non-Epic FHIR server is ever wired through this client and breaks the assumption, the typecheck failure is informative.

## Acceptance
- [ ] Decide: tighten to required, or keep optional + log
- [ ] If tightened, add a guard in `tryParseOperationOutcome` that rejects (returns undefined) Outcomes where any `coding` entry lacks `code`
- [ ] Test covering the rejected/logged case

## Refs
- PR #1100
- `services/epic-connector/src/fhir-types.ts`

### PR DESCRIPTION
**Pull request by blamechris**
## Summary
Bundles four PR #1100 follow-ups around the structured-error story. All touch `fhir-client.ts` / `fhir-types.ts` / `sync-jobs.ts` and would conflict if shipped separately.

- **#1101** — `operationOutcomeError` returns `EpicFhirError` instead of plain `Error`. FHIR-spec semantic-failure-on-200 (Epic does this for read/create/update edge cases) now flows through the structured-error pipeline; downstream `err.hasIssueCode("59022")` works for the 200-with-OO case the same way it works for 4xx-with-OO.
- **#1102** — New `epic-error-codes.ts` registry: `EPIC_ERROR_CODES.UNAUTHORIZED_SUB_RESOURCE` (59022) and `MISSING_REQUIRED_ELEMENT` (59108), each with policy docs. `sync-jobs.ts` swaps its inline `"59022"` literal for the registry symbol.
- **#1103** — `tryParseOperationOutcome` tightened with structural checks: rejects non-object JSON, missing/wrong `resourceType`, non-array `issue`. Five new direct edge-case tests lock the trust boundary in.
- **#1104** — `OperationOutcomeIssueCoding.code` tightened from optional to required. Parser rejects payloads with a coding entry missing `code`. Downstream code can trust that any promoted OperationOutcome has string codes throughout.

Closes #1101.
Closes #1102.
Closes #1103.
Closes #1104.

## Why bundle these
The four issues are facets of the same underlying problem: the structured-error story PR #1100 introduced was load-bearing but had four gaps. Fix any one in isolation and you still have the others. Bundled, the invariant becomes:

> Any \`EpicFhirError.operationOutcome\` that exists is structurally trustworthy — \`issue\` is an array, every \`coding[*]\` has a string \`code\`. Soft-skip / surface decisions can rely on \`hasIssueCode()\` instead of substring-matching error messages.

The substring fallback in `isUnauthorizedSubResourceError` stays — it's still needed for non-OperationOutcome edge cases (HTML 502 from a fronting proxy, plain-text errors from an edge layer where the OO body never made it through). The fallback is no longer covering for our own \"plain Error\" path though, which is the gap #1101 closed.

## API additions
- New module export: `EPIC_ERROR_CODES` + `EpicErrorCode` type (re-exported from package root)
- `OperationOutcomeIssueCoding.code` is now required — purely tightening the existing type, no runtime change
- `operationOutcomeError` is internal; the visible change is that `client.read/createResource/updateResource` against a 2xx-with-OO body now throws `EpicFhirError` instead of `Error`

## Test plan
- [x] `pnpm --filter @carebridge/epic-connector test` — 199 pass + 1 documented skip (was 187, +12 new)
  - 4 new EPIC_ERROR_CODES sanity tests
  - 5 new `tryParseOperationOutcome` rejection tests (string body, array body, null body, no resourceType, wrong resourceType)
  - 1 new `tryParseOperationOutcome` rejection test for coding-without-code (#1104)
  - 1 new acceptance test for coding-less OperationOutcome still being valid
  - 1 new acceptance test for 200-with-OO throwing EpicFhirError (#1101)
  - Updated existing \"issue: null\" test to reflect the new strict semantics (operationOutcome is now undefined, not defined)
- [x] `pnpm typecheck` — workspace 38/38 clean

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #1122

### Summary
Solid consolidation of the four #1100 follow-ups. The structured-error invariant ("any `EpicFhirError.operationOutcome` that exists is structurally trustworthy") is now actually enforceable, and the registry pattern matches the codebase convention (`as const` object, not enum). One real fidelity bug worth fixing in a follow-up — `operationOutcomeError` hardcodes status 200/OK even when called from `createResource` (typically 201). Approving on the merits; deferred items below.

### Strengths
- Trust-boundary tightening in `tryParseOperationOutcome` is clean: explicit `typeof !== "object" || === null || Array.isArray` rejects every non-plain-object case, and the coding-without-code loop is the right place to apply the #1104 type tightening at runtime.
- Edge-case parser tests correctly use 400 responses so they actually route through `requestWithRetry` → `tryParseOperationOutcome`. No test conflation between the 200-with-OO path (`operationOutcomeError`) and the 4xx-with-OO path (parser).
- `EPIC_ERROR_CODES` as `const` object with `as const` matches the codebase pattern (see `EPIC_SOURCE_SYSTEM_TAG`, `DEFAULT_OPTIONS`, `CAREBRIDGE_ORGANIZATION_REFERENCE`). No enums used elsewhere in the service.
- Doc-comments on each registry entry call out the policy (SOFT-SKIP vs SURFACE) right next to the code — exactly the right place for it.
- Substring-fallback in `isUnauthorizedSubResourceError` is explicitly justified for HTML-502 / plain-text edge-proxy cases. Defensible and documented.
- Updated "issue: null" test now correctly asserts `operationOutcome` is **undefined** — the previous assertion (it was *defined*) would have been a false-positive after the parser tightening.

### Issues Found

#### Critical (Must Fix)
(none)

#### Suggestions (Should Consider)
| File | Line | Suggestion | Rationale |
|------|------|------------|-----------|
| `services/epic-connector/src/fhir-client.ts` | 523-528 | `operationOutcomeError` hardcodes `status: 200, statusText: "OK"`. Plumb the real `response.status`/`response.statusText` from each of the three call sites (`read`/`createResource`/`updateResource`). | `createResource` typically returns **201 Created**; `updateResource` can be 200 or 201. After this PR, `err.status` is always 200 from this path — silently lying for write paths. The structured-error story still works (`hasIssueCode` reads `operationOutcome`), but downstream metrics/alerting that key on `err.status` get false data. Deferred to #1123. |
| `services/epic-connector/src/epic-error-codes.ts` | 39 | `MISSING_REQUIRED_ELEMENT` is defined but has no call site checking for it. Either wire up a `isMissingRequiredElementError` helper in `sync-jobs.ts` (matches the SURFACE policy in the doc-comment) or remove from registry until needed. | Registry drift risk: future devs assume the policy is enforced because the entry exists. Deferred to #1124. |
| `services/epic-connector/src/__tests__/fhir-client-write.test.ts` | 104-124 | Only `read`'s 200-with-OO path has a regression test for the #1101 `EpicFhirError` invariant. `createResource`/`updateResource` go through the same helper but the existing OO test only asserts `.rejects.toThrow()` — would still pass if a future refactor reverted `operationOutcomeError` to plain `Error`. | Lock in the #1101 invariant for all three call sites. Deferred to #1125. |

#### Nitpicks (Optional)
- Truncation constant `500` is duplicated in `safeReadBody` and `operationOutcomeError`. Extract as `const MAX_BODY_SNIPPET_BYTES = 500` so future changes stay in sync with the `EpicFhirError.body` JSDoc contract. Deferred to #1126.
- The inline cast in `tryParseOperationOutcome` (`for (const issue of candidate.issue as Array<{ details?: { coding?: Array<{ code?: unknown }> } }>)`) is a bit long. A named local type would read cleaner, but as-is is fine for a single-use cast.

### Deferred Items (Follow-Up Issues)
| Suggestion | Issue | Rationale for deferral |
|------------|-------|------------------------|
| Plumb real HTTP status into `operationOutcomeError` | #1123 | Touches three call sites and needs new write-path tests; cleaner as its own PR. |
| Wire up `MISSING_REQUIRED_ELEMENT` (59108) or remove from registry | #1124 | Touches `sync-jobs.ts` policy code, deserves its own design discussion. |
| Add `createResource`/`updateResource` 2xx-with-OO regression coverage | #1125 | Same surface as #1123; natural to bundle. |
| Extract `MAX_BODY_SNIPPET_BYTES` constant | #1126 | Pure DRY, no urgency. |

### Architecture Notes
The structured-error pipeline now has a clean trust boundary: raw Epic body → `tryParseOperationOutcome` (rejects non-promotable shapes) → typed `OperationOutcome` → `hasIssueCode`. The 4xx/5xx path (`requestWithRetry`) and the 2xx-semantic-failure path (`operationOutcomeError`) both terminate at `EpicFhirError`, so downstream consumers don't have to discriminate between them by `instanceof`. Substring fallback in `isUnauthorizedSubResourceError` is now confined to genuinely opaque payloads (HTML 502 from a fronting proxy) — the prior gap where it had to cover our own plain-Error path is closed. The registry pattern (`EPIC_ERROR_CODES` + `EpicErrorCode` type) is the right primitive to grow into as more codes get policy attached; consider tightening `hasIssueCode(code: string)` to `hasIssueCode(code: EpicErrorCode)` once the registry stabilises, but defer for now to avoid locking out ad-hoc codes in tests.

### Verdict
- [x] Approve
- [ ] Request Changes
- [ ] Comment

File: services/epic-connector/src/fhir-client.ts [hunk 1]
```diff
@@ -481,12 +505,27 @@ async function safeReadBody(response: Response): Promise<string> {
...
-  return new Error(
-    `Epic returned OperationOutcome: ${issue?.code ?? "unknown"} — ${detail}`,
+  const message = `Epic returned OperationOutcome: ${issue?.code ?? "unknown"} — ${detail}`;
+  return new EpicFhirError(
+    message,
+    200,
+    "OK",
+    JSON.stringify(outcome).slice(0, 500),
+    outcome,
   );
```
[the remark this answers is not in the document]
Comment (blamechris) [reply services/epic-connector/src/fhir-client.ts]: **FIX**

Fixed in `63bd94f` (the pragmatic half of #1123 pulled into this PR).

**Change:** `operationOutcomeError` now accepts `status` + `statusText` parameters and the three call sites pass their operation's expected success code:

```diff
- function operationOutcomeError(outcome: OperationOutcome): EpicFhirError {
+ function operationOutcomeError(
+   outcome: OperationOutcome,
+   status: number,
+   statusText: string,
+ ): EpicFhirError {
```

Call sites:
- `client.read()` → `200, "OK"`
- `client.createResource()` → `201, "Created"`
- `client.updateResource()` → `200, "OK"` (PUT typical-success)

This is still a per-caller default rather than the real wire status — plumbing the response object through `requestWithRetry` would be a wider refactor (tracked under #1123 if anyone needs the real status later). But it removes the misleading "always 200" on writes and makes log/dashboard data honest about operation context.

Two new regression tests lock the contextual statuses in:
- `client.createResource against a 2xx with OperationOutcome body throws EpicFhirError with status=201 Created`
- `client.updateResource against a 2xx with OperationOutcome body throws EpicFhirError with status=200 OK` (also verifies `MISSING_REQUIRED_ELEMENT` 59108 rides through `hasIssueCode` structurally)

Tests: 201 pass + 1 skip. Typecheck clean.

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | `operationOutcomeError` hardcodes status=200/statusText=OK; misreported for createResource (201) and updateResource (200/201) | FIX | `63bd94f` |

**Also rolled into the same commit (pulled forward from agent-review's deferred issues since the fix was small):**
- #1123 (real fidelity regression flagged by both Copilot and agent-review) — operationOutcomeError now takes status/statusText; call sites pass per-operation expected success codes (read=200, createResource=201, updateResource=200)
- #1125 (regression coverage for createResource/updateResource 2xx-with-OO) — 2 new tests assert the contextual statuses

**Total:** 1 inline + 2 pulled-forward → 1 fix commit
- Fixed: 1 (commit `63bd94f`)
- False positives: 0
- Deferred follow-up issues: 2 (#1124 — wire up or remove `EPIC_ERROR_CODES.MISSING_REQUIRED_ELEMENT` (now used in the test, but no hasIssueCode call yet); #1126 — extract `MAX_BODY_SNIPPET_BYTES` constant)
- Issues closed in this commit: 2 (#1123, #1125)

**Note on #1124:** the new updateResource regression test uses `EPIC_ERROR_CODES.MISSING_REQUIRED_ELEMENT` in its assertion (`expect(err.hasIssueCode("59108")).toBe(true)`), but no production call site checks for 59108 yet. The issue stays open until a real surface decides to discriminate on it.
